File: c/k_r_whitebook/exercise/htoi_2-3.c
/*white book*/

#include <stdio.h>
#include <ctype.h>
#include <string.h>

int htoi(char *ptr){
	int flag=0;
	int hasil=0;
	int base=1;
	char uji;

	while(*ptr != '\0'){
		uji = tolower(*ptr);
		if(isspace(uji))
			;
		else if(isalpha(uji) 
				&& (uji >= 'a' || uji <= 'f' || uji == 'x')){/* bagian huruf */
			if(uji == 'x')
				return -1;
			else{
				if(flag){
					hasil += (uji - 'a' + 10) * base;
					base*=16;
				}
				else
					return -1;
			}
		}
		else if(isdigit(uji)){		/* bagian angka 0 - 9 */
			if(uji == '0' && (tolower(*(ptr+1)) =='x')){
				if(flag)
					return -1;
				else{
					flag=1;
					ptr+=2;
					continue;
				}
			}
			if(flag){
				hasil += (uji - '0') * base;
				base*=16;
			}
			else
				return -1;

		}
		else
			return -1;		/* digit ilegal */
		ptr++;
	}
	return hasil;
}

/* percobaan */

int main(){
	char buffer[10];

	puts("masukan nilai hexai:");
	fgets(buffer, BUFSIZ - 1, stdin);
	buffer[strlen(buffer)-1] = '\0';
	if(htoi(buffer)!=-1)
		printf("nilai desimal dari %s = %d\n", buffer, htoi(buffer));
	else
		printf("%s memiliki format salah.\n", buffer);
	return 0;
}

File: c/k_r_whitebook/exercise/htoi_2-3.py
#! /usr/bin/python

number = str(raw_input("masukan angka:"));

def htoi(number):
    hasil=0;
    flag=0;
    base=1;
    uji='0';
    i=0;

    number = number.lower()
    while i < len(number):
        print("check char=",number[i],i)
        if(number[i].isspace()):
            pass
        elif(number[i].isalpha() and (number[i] >= 'a' or number[i]<='f'\
                or number[i] == 'x')):
            if(number[i] == 'x' and flag):
                return -1
            else:
                hasil += (ord(number[i]) - ord('a') + 10) * base;
                base*=16;
        elif(number[i].isdigit()):
            if(number[i] == '0' and number[i+1] == 'x'):
                if(flag):
                    return -1
                else:
                    print("dalam 0x checking")
                    flag=1
                    i+=2;
                    continue
            hasil += (ord(number[i]) - ord('0')) * base;
            base *=16
        else:
            return -1
        i+=1
    return hasil

# percobaan #
if(htoi(number)!= -1):
    print ("nilai desimal dari",number,"=",str(htoi(number)))
else:
    print(number,"memiliki format salah.")
